platform/api/azure/storage.go

```diff
@@ -16,98 +16,261 @@ package azure
-	url := fmt.Sprintf("%s/%s", azureImageURL, name)
-	response, err := a.client.SendAzureGetRequest(url)
+func tryCopyPageBlob(client *service.Client, containerName, blobName, sourceURL string) bool {
+	ctx := context.TODO()
+	srcPageBlobClient, err := pageblob.NewClientWithNoCredential(sourceURL, nil)
+	if err != nil {
+		return false
+	}
+	srcBlobClient := srcPageBlobClient.BlobClient()
+	srcProperties, err := srcBlobClient.GetProperties(ctx, nil)
+	if err != nil {
+		return false
+	}
+	if srcProperties.ContentLength == nil {
+		return false
+	}
+
+	// we have a size, so we can create the target container and pageblob in it
+	dstContainerClient := client.NewContainerClient(containerName)
+	_, err = dstContainerClient.Create(ctx, nil)
+	if err != nil && !bloberror.HasCode(err, bloberror.ContainerAlreadyExists, bloberror.ResourceAlreadyExists) {
+		return false
+	}
+	dstPageBlobClient := dstContainerClient.NewPageBlobClient(blobName)
+	_, err = dstPageBlobClient.Create(ctx, *srcProperties.ContentLength, nil)
 	if err != nil {
-		if management.IsResourceNotFoundError(err) {
-			return false, nil
+		return false
+	}
+	deleteBlob := true
+	defer func() {
+		if deleteBlob {
+			// ignore errors
+			_ = DeleteBlob(client, containerName, blobName)
 		}
+	}()
 
-		return false, err
+	// we have allocated a target blob, now copy the pages
+	var marker *string
+	for {
+		opts := pageblob.GetPageRangesOptions{
+			Marker: marker,
+		}
+		pager := srcPageBlobClient.NewGetPageRangesPager(&opts)
+		for pager.More() {
+			response, err := pager.NextPage(ctx)
+			if err != nil {
+				return false
+			}
+			for _, page := range response.PageRange {
+				offset := *page.Start
+				// Both the page start and end are
+				// inclusive, thus the "+ 1" to obtain
+				// the length.
+				count := *page.End - *page.Start + 1
+				_, err = dstPageBlobClient.UploadPagesFromURL(ctx, sourceURL, offset, offset, count, nil)
+				if err != nil {
+					return false
+				}
+			}
+			marker = response.NextMarker
+		}
+		if marker == nil || *marker == "" {
+			break
+		}
 	}
 
-	var image OSImage
+	deleteBlob = false
+	return true
+}
 
-	if err := xml.Unmarshal(response, &image); err != nil {
-		return false, err
+func tryCopyAzcopy(client *service.Client, containerName, blobName, sourceURL string) bool {
+	azcopy, err := exec.LookPath("azcopy")
+	if err != nil {
+		return false
 	}
 
-	if image.Name == name {
-		return true, nil
+	blobClient := client.NewContainerClient(containerName).NewBlobClient(blobName)
+	sasPermissions := sas.BlobPermissions{}
+	sasPermissions.Read = true
+	sasPermissions.Write = true
+	sasPermissions.Create = true
+	expiry := time.Now().Add(15 * time.Minute)
+	dstSAS, err := blobClient.GetSASURL(sasPermissions, expiry, nil)
+	if err != nil {
+		return false
 	}
-
-	return false, nil
+	// log-level=NONE only affects the log file - stdout is unaffected
+	cmd := exec.Command(azcopy, "cp", "--blob-type=PageBlob", "--log-level=NONE", sourceURL, dstSAS)
```

Comment:
there is no lib for azcopy, and the the SDK CopyBlob function copies all 30GB of a sparse VHD instead of just the populated 1GB. But I think this code path is unused since we switched to the container pipeline (we don't copyblobs anymore, we always upload a vhd).

Reply:
`azcopy` is doing similar thing to the `tryCopyPageBlob` function, but with goroutines and throttling and stuff. What I wasn't sure about is whether to keep the code that invokes `azcopy`.